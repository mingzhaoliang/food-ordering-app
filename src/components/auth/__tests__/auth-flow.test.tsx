import { act, render, screen } from "@testing-library/react";
import AuthFlow from "../auth-flow";

vi.mock("@/actions/auth/signIn");
vi.mock("@/actions/auth/signUp");

vi.mock("next/headers", () => ({
  headers: vi.fn(() => ({
    referer: "http://localhost:3000/",

    get() {
      return this.referer;
    },
  })),
}));

describe("AuthFlow", () => {
  ["signin", "signup"].forEach((mode) => {
    it(`should match the snapshot in ${mode} mode`, async () => {
      const searchParams = { mode };
      const { asFragment } = render(<AuthFlow searchParams={searchParams} />);
      await act(async () => {
        await vi.dynamicImportSettled();
      });

      expect(asFragment()).toMatchSnapshot();
    });
  });

  it("should render the signin form in signin mode", async () => {
    const searchParams = { mode: "signin" };
    render(<AuthFlow searchParams={searchParams} />);
    await act(async () => {
      await vi.dynamicImportSettled();
    });

    expect(screen.getByRole("form")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Sign in" })).toBeInTheDocument();
    expect(screen.getByText("Don't have an account?")).toBeInTheDocument();
    expect(screen.getByText("Create an account")).toBeInTheDocument();
  });

  it("should render the google button in signin mode", () => {
    const searchParams = { mode: "signin" };
    render(<AuthFlow searchParams={searchParams} />);

    expect(screen.getByText("Continue with Google")).toBeInTheDocument();
  });

  it("should render the google button in signup mode", () => {
    const searchParams = { mode: "signup" };
    render(<AuthFlow searchParams={searchParams} />);

    expect(screen.getByText("Continue with Google")).toBeInTheDocument();
  });

  it("should render the signup form in signup mode", async () => {
    const searchParams = { mode: "signup" };
    render(<AuthFlow searchParams={searchParams} />);
    await act(async () => {
      await vi.dynamicImportSettled();
    });

    expect(screen.getByRole("form")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Sign up" })).toBeInTheDocument();
    expect(screen.getByText("Already have an account?")).toBeInTheDocument();
    expect(screen.getByText("Sign in")).toBeInTheDocument();
  });
});
