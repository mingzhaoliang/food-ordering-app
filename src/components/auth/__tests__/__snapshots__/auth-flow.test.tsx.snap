// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthFlow > should match the snapshot in signin mode 1`] = `
<DocumentFragment>
  <div
    class="flex-col-center gap-2 max-xs:text-sm"
  >
    <form
      action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
      class="font-lato flex flex-col w-full gap-1"
      role="form"
    >
      <div
        class="space-y-2"
      >
        <div
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="relative flex items-center"
          id=":r0:-form-item"
        >
          <div
            class="flex-1"
          >
            <input
              class="flex h-10 w-full border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-full xl:p-5 xl:text-base xs:h-12"
              data-testid="email"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
        </div>
        <p
          class="font-medium text-destructive text-xs pl-1"
          data-testid="form-message-email"
          id=":r0:-form-item-message"
        >
          <span
            class="inline-block h-3"
          />
        </p>
      </div>
      <div
        class="space-y-2"
      >
        <div
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="relative flex items-center"
          id=":r1:-form-item"
        >
          <div
            class="flex-1"
          >
            <input
              class="flex h-10 w-full border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-full xl:p-5 xl:text-base xs:h-12 pr-10"
              data-testid="password"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
        </div>
        <p
          class="font-medium text-destructive text-xs pl-1"
          data-testid="form-message-password"
          id=":r1:-form-item-message"
        >
          <span
            class="inline-block h-3"
          />
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-active disabled:hover:bg-primary h-10 px-4 py-2 2xl:h-11 mt-2"
        type="submit"
      >
        Sign in
      </button>
    </form>
    <div
      class="relative w-full flex py-3 items-center"
    >
      <div
        class="flex-grow border-t border-stone-300"
      />
      <p
        class="px-4 text-stone-500"
      >
        or
      </p>
      <div
        class="flex-grow border-t border-stone-300"
      />
    </div>
    <div
      class="flex w-full flex-col gap-2"
    >
      <a
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 2xl:h-11"
        href="/api/auth/signin/google?callbackUrl=http:/localhost:3000/"
      >
        <img
          alt="Google"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/src/assets/icons/google-g.svg"
          style="color: transparent;"
          width="30"
        />
        Continue with Google
      </a>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 2xl:h-11"
      >
        Continue with a demo account
      </button>
    </div>
    <div
      class="pt-4 text-sm text-stone-500"
    >
      Don't have an account? 
      <a
        class="text-teal-600 hover:underline"
        href="/auth?mode=signup"
      >
        Create an account
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AuthFlow > should match the snapshot in signup mode 1`] = `
<DocumentFragment>
  <div
    class="flex-col-center gap-2 max-xs:text-sm"
  >
    <form
      action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
      class="font-lato flex flex-col w-full gap-1"
      role="form"
    >
      <div
        class="grid max-lg:grid-cols-1 grid-cols-2 gap-x-4 gap-y-1"
      >
        <div
          class="space-y-2"
        >
          <div
            aria-describedby=":r2:-form-item-description"
            aria-invalid="false"
            class="relative flex items-center"
            id=":r2:-form-item"
          >
            <div
              class="flex-1"
            >
              <input
                class="flex h-10 w-full border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-full xl:p-5 xl:text-base xs:h-12"
                data-testid="firstName"
                id="firstName"
                name="firstName"
                placeholder="First name"
                value=""
              />
            </div>
          </div>
          <p
            class="font-medium text-destructive text-xs pl-1"
            data-testid="form-message-firstName"
            id=":r2:-form-item-message"
          >
            <span
              class="inline-block h-3"
            />
          </p>
        </div>
        <div
          class="space-y-2"
        >
          <div
            aria-describedby=":r3:-form-item-description"
            aria-invalid="false"
            class="relative flex items-center"
            id=":r3:-form-item"
          >
            <div
              class="flex-1"
            >
              <input
                class="flex h-10 w-full border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-full xl:p-5 xl:text-base xs:h-12"
                data-testid="lastName"
                id="lastName"
                name="lastName"
                placeholder="Last name (optional)"
                value=""
              />
            </div>
          </div>
          <p
            class="font-medium text-destructive text-xs pl-1"
            data-testid="form-message-lastName"
            id=":r3:-form-item-message"
          >
            <span
              class="inline-block h-3"
            />
          </p>
        </div>
      </div>
      <div
        class="space-y-2"
      >
        <div
          aria-describedby=":r4:-form-item-description"
          aria-invalid="false"
          class="relative flex items-center"
          id=":r4:-form-item"
        >
          <div
            class="flex-1"
          >
            <input
              class="flex h-10 w-full border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-full xl:p-5 xl:text-base xs:h-12"
              data-testid="email"
              id="email"
              name="email"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
        </div>
        <p
          class="font-medium text-destructive text-xs pl-1"
          data-testid="form-message-email"
          id=":r4:-form-item-message"
        >
          <span
            class="inline-block h-3"
          />
        </p>
      </div>
      <div
        class="space-y-2"
      >
        <div
          aria-describedby=":r5:-form-item-description"
          aria-invalid="false"
          class="relative flex items-center"
          id=":r5:-form-item"
        >
          <div
            class="flex-1"
          >
            <input
              class="flex h-10 w-full border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 rounded-full xl:p-5 xl:text-base xs:h-12"
              data-testid="password"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="absolute right-4"
          >
            <a
              data-state="closed"
            >
              <svg
                class="lucide lucide-info w-5 h-5 text-fade"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <circle
                  cx="12"
                  cy="12"
                  r="10"
                />
                <path
                  d="M12 16v-4"
                />
                <path
                  d="M12 8h.01"
                />
              </svg>
            </a>
          </div>
        </div>
        <p
          class="font-medium text-destructive text-xs pl-1"
          data-testid="form-message-password"
          id=":r5:-form-item-message"
        >
          <span
            class="inline-block h-3"
          />
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-active disabled:hover:bg-primary h-10 px-4 py-2 2xl:h-11 xs:h-12 mt-2 xl:text-base"
        type="submit"
      >
        Sign up
      </button>
    </form>
    <div
      class="relative w-full flex py-3 items-center"
    >
      <div
        class="flex-grow border-t border-stone-300"
      />
      <p
        class="px-4 text-stone-500"
      >
        or
      </p>
      <div
        class="flex-grow border-t border-stone-300"
      />
    </div>
    <div
      class="flex w-full flex-col gap-2"
    >
      <a
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 2xl:h-11"
        href="/api/auth/signin/google?callbackUrl=http:/localhost:3000/"
      >
        <img
          alt="Google"
          data-nimg="1"
          decoding="async"
          height="30"
          loading="lazy"
          src="/src/assets/icons/google-g.svg"
          style="color: transparent;"
          width="30"
        />
        Continue with Google
      </a>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-full text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 2xl:h-11"
      >
        Continue with a demo account
      </button>
    </div>
    <div
      class="pt-4 text-sm text-stone-500"
    >
      Already have an account? 
      <a
        class="text-teal-600 hover:underline"
        href="/auth?mode=signin"
      >
        Sign in
      </a>
    </div>
  </div>
</DocumentFragment>
`;
